clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java

```diff
@@ -149,15 +149,17 @@ public HeartbeatRequestManager(
      */
     @Override
     public NetworkClientDelegate.PollResult poll(long currentTimeMs) {
-        if (!coordinatorRequestManager.coordinator().isPresent() || !membershipManager.shouldSendHeartbeat())
+        if (!coordinatorRequestManager.coordinator().isPresent() || membershipManager.shouldSkipHeartbeat())
             return NetworkClientDelegate.PollResult.EMPTY;
 
-        // TODO: We will need to send a heartbeat response after partitions being revoke. This needs to be
-        //  implemented either with or after the partition reconciliation logic.
-        if (!heartbeatRequestState.canSendRequest(currentTimeMs))
+        boolean heartbeatNow = membershipManager.shouldHeartbeatNow() && !heartbeatRequestState.requestInFlight();
+
+        if (!heartbeatRequestState.canSendRequest(currentTimeMs) && !heartbeatNow) {
             return new NetworkClientDelegate.PollResult(heartbeatRequestState.nextHeartbeatMs(currentTimeMs));
+        }
 
-        this.heartbeatRequestState.onSendAttempt(currentTimeMs);
+        heartbeatRequestState.onSendAttempt(currentTimeMs);
+        membershipManager.onHeartbeatRequestSent();
```

Comment:
I'm using `send` to align with how the HB manager was naming the existing `onSendAttempt` and such, and I think that as seen from the manager point of view it is enough, the membership manager only needs to know that the request is sent out of the HB manager to transition accordingly. It is not literally sent over the network. I would leave send/sent to keep it consistent between the 2 managers, but let me know if you think it would be clearer differently.